### PR TITLE
fix #6745 feat(nimbus): Use a multi-select for feature ids

### LINKED ISSUE #6745: Allow selecting multiple features and inputting multiple values for them in nimbus-ui
After the V5 API has been migrated to use the new `feature_configs` and `feature_values` fields we can update the UI to allow users to input multiple features and their values in each branch.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/EXP-1919)

### PR DESCRIPTION
**Pull request by brennie**
Because

- we are moving to a multi-feature world;
- multifeature support has not been QA'd on all platforms yet (see
  [QA-1977][1]); and

this commit

- updates the branches page to use a multi-select for features; and
- limits the selection to a single feature, which can be undone once QA
  has finished.

[1]: https://mozilla-hub.atlassian.net/browse/QA-1977

### CODE REVIEW

**Comment by brennie** [comment]
Blocked on #8860

**Review by jaredlockhart — COMMENTED** [review]
Got into a couple interesting states by triggering some errors, then adding and removing features, the list of values gets out of sync with the list of errors in memory.  

image

Refresh page:

image


I think we have to assert some kind of stable ordering here, like sorting by feature slug?  That way we don't get the values and errors lists out of sync?  Or maybe when the features list changes we should just throw out the existing errors?

**Comment by jaredlockhart** [comment]
Okay!  Testing locally things are looking really goooooodddd

I noticed the summary page doesn't show the feature names:

image

Do you wanna handle that in here or do it as a followup?

File: experimenter/experimenter/nimbus-ui/src/components/Summary/TableBranches/index.tsx [hunk 1]
```diff
@@ -190,12 +190,23 @@ const TableBranch = ({
...
-        <tr>
-          <th>Value</th>
-          <td colSpan={3} data-testid="branch-featureValue">
-            {featureValue ? <Code codeString={featureValue} /> : <NotSet />}
-          </td>
-        </tr>
+        {featureValues ? (
+          featureValues.map((featureValue, idx) => (
+            <tr key={idx}>
+              <th>Value</th>
```
Comment (jaredlockhart) [inline experimenter/experimenter/nimbus-ui/src/components/Summary/TableBranches/index.tsx:196]: Right here could just be

<feature name> Value

Comment (jaredlockhart) [reply experimenter/experimenter/nimbus-ui/src/components/Summary/TableBranches/index.tsx:196]: And probably stable ordered by feature name/slug?

**Comment by brennie** [comment]
I updated the API to always return the feature values ordered by the feature config ID. I also updated the frontend select component and input values to be ordered the same.

**Comment by brennie** [comment]
I updated the v5 serializer to ensure that it always orders values by their feature config IDs, just like the graphQL api.

File: experimenter/experimenter/experiments/api/v5/serializers.py [hunk 2]
```diff
@@ -379,6 +379,21 @@ class Meta:
...
+
+        This enforces that the serialized data is ordered by the feature config ID.
+        """
+        super().to_representation
+        iterable = data.all() if isinstance(data, models.Manager) else data
+
+        return [
+            self.child.to_representation(item)
+            for item in iterable.order_by("feature_config__id")
+        ]
```
Comment (jaredlockhart) [inline experimenter/experimenter/experiments/api/v5/serializers.py:394]: Yeah exactly.  Not obvious, kind of a hassle.  But yep.

File: experimenter/experimenter/experiments/tests/api/v5/test_serializers/test_nimbus_experiment_serializer.py [hunk 3]
```diff
@@ -327,15 +327,25 @@ def test_saves_branches_single_feature(self):
...
                     "ratio": 1,
-                    "feature_value": "{'this': 'that'}",
-                }
+                    "feature_values": [
+                        {
+                            "feature_config": feature_config.id,
+                            "value": "{'this': 'that'}",
+                        }
+                    ],
+                },
```
Comment (jaredlockhart) [inline experimenter/experimenter/experiments/tests/api/v5/test_serializers/test_nimbus_experiment_serializer.py:348]: Might be worth putting two features in there and asserting their presence and ordering?

**Review by jaredlockhart — APPROVED** [review]
Okay wow.  Can't believe it's finally here.  Tested locally, everything was A++ 11/10.  Great job.  Form works fantastically, summary page has names, recipe serializer includes the values properly.  I think it all looks right.  Only two things jumping out at me are
* Could cover the feature value ordering in a test somewhere
* Still think we might wanna nuke the warnings/errors if the feature config selector changes

But nothing blocking and things we can do as followups if we want.

Thank you @brennie this is gonna be huge 🎉 🎉 🎉 🎉 🎉

**Review by jaredlockhart — DISMISSED** [review]
Oh now I'm not getting any errors?

image


image

After clicking save.

Refreshing the page brings them back

image


Errors/warnings should be shown immediately after clicking save.

**Review by jaredlockhart — APPROVED** [review]
Okay.  Ran through it all again.  THIS is the best experience it's been so far.  I never got into a confused state.  Resetting the errors made a big difference.  Now the UI state is always consistent.  All the validation is all working in all directions, every combination I could think of.  Thank you @brennie for getting this over the finish line.  Now let's land this thing!! 🎉 🎉 🎉 🎉

**Comment by yashikakhurana** [comment]
wooo hoooo can't wait to see this landing 🎉 , great job  @brennie 👏

**Comment by brennie** [comment]
Blocked on QA

**Comment by jaredlockhart** [comment]
Let's block on this https://mozilla-hub.atlassian.net/browse/EXP-3608

**Comment by brennie** [comment]
Update: I've removed the parts of this that change the API responses and have limited the frontend to only allow selecting one feature at a time.

**Comment by brennie** [comment]
https://github.com/mozilla/experimenter/pull/8907/files#diff-99544655f6d0c765eebe36ad77f4df48ff10ab000d88ac14a84e25645aef56f6R208-R209 contains the logic for limiting the frontend to a single value. It is accompanied by a test. Removal of that line will allow for multi-feature frontend.

File: experimenter/experimenter/nimbus-ui/src/components/PageEditBranches/FormBranches/index.tsx [hunk 4]
```diff
@@ -176,11 +201,15 @@ export const FormBranches = ({
...
 
-  const onFeatureConfigChange = (ev: React.ChangeEvent<HTMLInputElement>) => {
-    const selectedFeatureId = parseInt(ev.target.value, 10);
-    setIsSelectValid(true);
+  const onFeatureConfigsChanged = (newValue: Options<FeatureConfigOption>) => {
+    setIsSelectValid(newValue.length > 0);
+    setSelectDirty(true);
+
+    // Limit to a single value until multi-feature support has been QA'd.
+    newValue = newValue.length > 0 ? [newValue[newValue.length - 1]] : newValue;
```
Comment (jaredlockhart) [inline experimenter/experimenter/nimbus-ui/src/components/PageEditBranches/FormBranches/index.tsx:209]: 👍

**Review by jaredlockhart — APPROVED** [review]
Okay yep, tested locally, everything looks right.  Funny, always chop off the selected list to the last new entry.

I was thinking of this like on the metrics page:

https://github.com/mozilla/experimenter/blob/main/experimenter/experimenter/nimbus-ui/src/components/PageEditMetrics/FormMetrics/index.tsx#L163

But this works too.

Awesome let's get this landed 🎉 🎉 🎉

**Pull request merged**